caches/infinispan/src/main/java/org/commonjava/maven/galley/cache/infinispan/FastLocalCacheProvider.java

```diff
@@ -0,0 +1,623 @@
+    // in NFS. Used as <path, ip> cache to collect nfs ownership of the file storage
+    @ConfigureCache( "nfs-owner-cache" )
+    @NFSOwnerCache
+    private Cache<String, String> nfsOwnerCache;
+
+    private TransactionManager cacheTxMgr;
+
+    @Inject
+    private FileEventManager fileEventManager;
+
+    @Inject
+    private TransferDecorator transferDecorator;
+
+    @ExecutorConfig( named = "fast-local-executor", threads = 5, priority = 2, daemon = true )
+    @WeftManaged
+    @Inject
+    private Executor executor;
+
+    protected FastLocalCacheProvider()
+    {
+        nfsBaseDir = System.getProperty( NFS_BASE_DIR_KEY );
+    }
+
+    public FastLocalCacheProvider( final PartyLineCacheProvider plCacheProvider,
+                                   final Cache<String, String> nfsUsageCache, final FileEventManager fileEventManager,
+                                   final TransferDecorator transferDecorator, final Executor executor )
+    {
+        this.plCacheProvider = plCacheProvider;
+        this.nfsOwnerCache = nfsUsageCache;
+        this.fileEventManager = fileEventManager;
+        this.transferDecorator = transferDecorator;
+        this.executor = executor;
+    }
+
+    public FastLocalCacheProvider( final PartyLineCacheProvider plCacheProvider,
+                                   final Cache<String, String> nfsUsageCache, final FileEventManager fileEventManager,
+                                   final TransferDecorator transferDecorator, final Executor executor,
+                                   final String nfsBaseDir )
+    {
+        this( plCacheProvider, nfsUsageCache, fileEventManager, transferDecorator, executor );
+        this.nfsBaseDir = nfsBaseDir;
+    }
+
+    @PostConstruct
+    public void init()
+    {
+        cacheTxMgr = nfsOwnerCache.getAdvancedCache().getTransactionManager();
+
+        startReporting();
+    }
+
+    @PreDestroy
+    public void destroy()
+    {
+        stopReporting();
+    }
+
+    @Override
+    public boolean isFileBased()
+    {
+        return true;
+    }
+
+    @Override
+    public File getDetachedFile( ConcreteResource resource )
+    {
+        //This Detached file is local one first, if not found will use nfs one.
+        File file = plCacheProvider.getDetachedFile( resource );
+        if ( file == null && StringUtils.isNotBlank( nfsBaseDir ) )
+        {
+            file = getNFSDetachedFile( resource );
+        }
+
+        return file;
+    }
+
+    private File getNFSDetachedFile( ConcreteResource resource )
+    {
+        String path = PathUtils.normalize( nfsBaseDir, getFilePath( resource ) );
+        return new File( path );
+    }
+
+    @Override
+    public void startReporting()
+    {
+        plCacheProvider.startReporting();
+    }
+
+    @Override
+    public void stopReporting()
+    {
+        plCacheProvider.stopReporting();
+    }
+
+    @Override
+    public void cleanupCurrentThread()
+    {
+        plCacheProvider.cleanupCurrentThread();
```

Comment:
WRT expiring streams, I think this is actually a good, reliable way to do it. We should keep a map `Map<threadId, Set<OutputStream>>` in this cache provider. When we open a new output stream wrapper to pass to the user, we should add it to the map using `Thread.currentThread().getId()` for the key. When this method (`cleanupCurrentThread()`) gets called, we should again grab the current thread's id, remove the corresponding set from the map, and iterate through it, closing each stream.

This means openOutputStream() and cleanupCurrentThread() will either have to be `synchronized` (or contain `synchronized` blocks), or else we might be able to get away with a ConcurrentHashMap...though if I remember correctly, that may be a pretty memory-hungry class.

The idea here is that after the user is done with a thread, some cleanup mechanism will run. In Indy, this happens in a servlet Filter, in a finally block (IIRC). All opportunity for user interaction is past, and the system cleans up any dangling resources related to that request.

For an example, see:

* https://github.com/Commonjava/partyline/blob/master/src/main/java/org/commonjava/util/partyline/JoinableFileManager.java#L67
* https://github.com/Commonjava/indy/blob/master/subsys/jaxrs/src/main/java/org/commonjava/indy/bind/jaxrs/ResourceManagementFilter.java#L71